File: csharp/UrakawaToolkit/property/xml/IXmlPropertyFactory.cs
using System;
using System.Collections.Generic;
using System.Text;
using urakawa.property;

namespace urakawa.property.xml
{
	/// <summary>
	/// <see cref="PropertyFactory"/> that supports creation of <see cref="XmlProperty"/>s 
	/// and <see cref="XmlAttribute"/>s
	/// </summary>
	public interface IXmlPropertyFactory : IGenericPropertyFactory
	{
		/// <summary>
		/// Creates a <see cref="XmlProperty"/> of default type
		/// </summary>
		/// <returns>The created <see cref="XmlProperty"/></returns>
		XmlProperty createXmlProperty();

		/// <summary>
		/// Creates a <see cref="XmlAttribute"/> of default type
		/// with a given parent <see cref="XmlProperty"/>
		/// </summary>
		/// <param name="parent">The parent <see cref="XmlProperty"/></param>
		/// <returns>The created <see cref="XmlAttribute"/></returns>
		XmlAttribute createXmlAttribute(XmlProperty parent);

		/// <summary>
		/// Creates a <see cref="XmlAttribute"/> of type matching a given QName 
		/// with a given parent <see cref="XmlProperty"/>
		/// </summary>
		/// <param name="localName">The local localName part of the given QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given QName</param>
		/// <param name="parent">The parent <see cref="XmlProperty"/></param>
		/// <returns>The created <see cref="XmlAttribute"/>, <c>null</c> if the given QName is not recognized</returns>
		XmlAttribute createXmlAttribute(XmlProperty parent, string localName, string namespaceUri);
	}
}

File: trunk/java/core/src/org/daisy/urakawa/property/xml/IXmlPropertyFactory.java
package org.daisy.urakawa.property.xml;

import org.daisy.urakawa.IWithPresentation;

import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * <p>
 * This is the factory that creates
 * {@link org.daisy.urakawa.property.xml.IXmlProperty} and
 * {@link org.daisy.urakawa.property.xml.IXmlAttribute} instances.
 * </p>
 * 
 * @see org.daisy.urakawa.property.PropertyFactory
 * @designConvenienceInterface see
 *                             {@link org.daisy.urakawa.DesignConvenienceInterface}
 * @see org.daisy.urakawa.DesignConvenienceInterface
 * @stereotype OptionalDesignConvenienceInterface
 * @depend - Create - IXmlProperty
 * @depend - Create - IXmlAttribute
 */
public interface IXmlPropertyFactory extends IXukAble, IWithPresentation {
	/**
	 * <p>
	 * Creates a new property, not yet associated to a node.
	 * </p>
	 * <p>
	 * This factory method does not take any argument and creates an object of
	 * the default type.
	 * </p>
	 * 
	 * @return cannot be null.
	 */
	public IXmlProperty createXmlProperty();

	/**
	 * <p>
	 * Creates a new XML attribute.
	 * </p>
	 * 
	 * @return cannot be null.
	 */
	public IXmlAttribute createXmlAttribute();

	/**
	 * <p>
	 * Creates a new XML attribute.
	 * </p>
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * 
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 */
	public IXmlAttribute createXmlAttribute(String xukLocalName,
			String xukNamespaceURI) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;
}
